Docs: merge REST API query instructions

ltalirz opened the pull request:
fixes #4453 

Merge the instructions for querying the REST API into the new "How to
share data" section; moving the details on pagination into the REST API
reference.

Note:  I should probably mention that

 *  I haven't deleted the `Computer` section of the REST API reference - I just moved it down in the list in order to start the list with `Nodes`, which is likely the most important resource by far.
 * in the REST API reference, you will find a couple of `|Group| s` (i.e. an extra space between the class reference and the plural-s). I put this in because `|Group|s` confuses sphinx (with or without the `|...|` substitution syntax) - happy to fix this if there is a fix

giovannipizzi commented:
I didn't review this since I see @CasperWA  is assigned, but please remember to add a redirect before  merging

CasperWA commented on docs/source/howto/share_data.rst:
```diff
@@ -155,9 +155,45 @@ Like all ``verdi`` commands, you can select a different AiiDA profile via the ``
+A URL to query the REST API consists of:
 
-For an extensive user documentation, see the :ref:`AiiDA REST API reference <reference:rest-api>`.
+
```
```suggestion
```

CasperWA commented on docs/source/howto/share_data.rst:
```diff
@@ -155,9 +155,45 @@ Like all ``verdi`` commands, you can select a different AiiDA profile via the ``
 
-For an extensive user documentation, see the :ref:`AiiDA REST API reference <reference:rest-api>`.
+
+    1. The *base URL*, by default:
+
+        http://127.0.0.1:5000/api/v4
+
+        Querying the base URL returns a list of all available endpoints.
+
+    2. The *path* defining the requested *resource*, optionally followed by a more specific *endpoint*. For example::
+
+            /nodes
+            /nodes/page/2
+            /nodes/projectable_properties
+            /nodes/<uuid>
+            /nodes/<uuid>/links/outgoing
+
+       If no endpoint is appended, the API returns a list of objects of that resource.
+       In order to request a specific object of a resource, append its *UUID*.
+
+       .. note::
+
+           As usual, you can use partial UUIDs as long as they are unique.
+
+           In order to query by *PK* you need to use the ``id`` filter (see below).
+           This also applies to :py:class:`~aiida.orm.User` s, which don't have UUIDs.
+
+
+    3. (Optional) The *query string* for filtering, ordering and pagination of results. For example::
+
+        ?limit=20&offset=35
+        ?id=200
+        ?node_type=like="data%"
```
This whole section should be de-indented. Also check the indentation of the parts, I believe the line `Querying the base URL ...` is prepended with one space too many?

ltalirz replied:
Thanks for noticing, will fix!

CasperWA commented on docs/source/howto/share_data.rst:
```diff
@@ -155,9 +155,45 @@ Like all ``verdi`` commands, you can select a different AiiDA profile via the ``
+           In order to query by *PK* you need to use the ``id`` filter (see below).
+           This also applies to :py:class:`~aiida.orm.User` s, which don't have UUIDs.
+
+
```
```suggestion
```

CasperWA commented on docs/source/howto/share_data.rst:
```diff
@@ -155,9 +155,45 @@ Like all ``verdi`` commands, you can select a different AiiDA profile via the ``
+           As usual, you can use partial UUIDs as long as they are unique.
+
+           In order to query by *PK* you need to use the ``id`` filter (see below).
+           This also applies to :py:class:`~aiida.orm.User` s, which don't have UUIDs.
```
```suggestion
           This also applies to :py:class:`~aiida.orm.users.User` s, which don't have UUIDs (but instead uses email).
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -34,109 +38,18 @@ The JSON object further contains information on the request in the ``method``, `
     Description:
 
-        returns the list of two *Node* objects (``limit=2``) starting from 9th row (``offset=8``) of the database table and the list will be ordered by ``id`` in descending order.
+        returns the list of two |Node| objects (``limit=2``) starting from 9th row (``offset=8``) of the database table and the list will be ordered by ``id`` in descending order.
```
Won't the indentation here make this a comment or similar?

ltalirz replied:
seems to work fine
![image](https://user-images.githubusercontent.com/1053245/96650601-70b37100-1333-11eb-9bd8-5e0b911e8cb1.png)


Admittedly, this is not the most beautiful formatting ever conceived but I think it's good enough to be understandable.

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -184,7 +97,7 @@ Nodes
     Description:
 
-        returns the list of *Node* objects. Every node object contains value of attribute called ``pbc1`` if present otherwise null.
+        returns the list of |Node| objects. Every node object contains value of attribute called ``pbc1`` if present otherwise null.
```
```suggestion
        returns the list of |Node| objects.
        Every node object contains value of attribute called ``pbc1`` if present otherwise null.
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -227,7 +140,7 @@ Nodes
         }
 
-2. Get a list of all available node types from database.
+2. Get a list of all available |Node| types from database.
```
```suggestion
2. Get a list of all available |Node| types from the database.
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -758,10 +671,102 @@ CalcJobs
+
+    Description:
+
+        returns the list of three |Computer| objects (``limit=3``) starting from the 3rd row (``offset=2``) of the database table and the list will be ordered by ascending values of ``id``.
```
```suggestion
        returns the list of three |Computer| objects (``limit=3``) starting from the 3rd row (``offset=2``) of the database table.
        The list will be ordered by ascending values of ``id``.
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -913,30 +918,62 @@ Groups
+If no page number is specified, the system redirects the request to page 1.
+When pagination is used, the **header** of the response contains two more non-empty fields:
+
+    - ``X-Total-Counts`` (custom field): the total number of results returned by the query, i.e. the sum of the results
+      of all pages
```
```suggestion
    - ``X-Total-Counts`` (custom field): the total number of results returned by the query, i.e. the sum of the results of all pages.
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -913,30 +918,62 @@ Groups
+
+    - ``X-Total-Counts`` (custom field): the total number of results returned by the query, i.e. the sum of the results
+      of all pages
+    - ``Links``: links to the first, previous, next, and last page. Suppose that you send a request whose results  fill
+      8 pages. Then the value of the ``Links`` field would look like::
```
```suggestion
    - ``Links``: links to the first, previous, next, and last page. Suppose that you send a request whose results fill 8 pages.
      Then the value of the ``Links`` field would look like::
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -913,30 +918,62 @@ Groups
+    :offset: Skips the first ``offset`` results (integer).
 
-    :perpage: Same format as ``limit``.
+    :perpage: How many results to show per page when paginating (integer).
```
I guess we're always "paginating"? :)
```suggestion
    :perpage: How many results to show per page (integer).
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -913,30 +918,62 @@ Groups
-        Additionally, ``+`` or ``-`` can be pre-pended to the value in order to select, respectively, ascending or descending order.
-        Specifying no leading character is equivalent to select ascending order.
+    :orderby: ``+<property>`` for ascending order and ``-<property>`` for descending order (``<property`` defaults to ascending).
         Ascending (descending) order for strings corresponds to alphabetical (reverse-alphabetical) order, whereas for datetime objects it corresponds to chronological (reverse-chronological order).
```
```suggestion
        Ascending (descending) order for strings corresponds to alphabetical (reverse-alphabetical) order, whereas for datetime objects it corresponds to chronological (reverse-chronological) order.
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -959,12 +996,10 @@ All of them must be followed by the operator ``=``.
     :attributes:
-        by default ``attributes`` are not returned in ``/nodes`` endpoint.
-        To get the list of all ``attributes`` specify ``attributes=true`` and to get selected ``attribute(s)`` list, use ``attributes=true&attributes_filters=<comma separated list of attributes you want to request>``.
+        Pass ``true`` in order to return attributes in the ``/nodes`` endpoint.
```
```suggestion
        Pass ``true`` in order to return attributes in the ``/nodes`` endpoint (excluded by default).
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -959,12 +996,10 @@ All of them must be followed by the operator ``=``.
     :extras:
-        by default ``extras`` are not returned in ``/nodes`` endpoint.
-        To get the list of all ``extras`` specify ``extras=true`` and to get selected ``extras`` list, use ``extras=true&extras_filters=<comma separated list of extras you want to request>``.
+        Pass ``true`` in order to return extras in the ``/nodes`` endpoint.
```
```suggestion
        Pass ``true`` in order to return extras in the ``/nodes`` endpoint (excluded by default).
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -1142,3 +1177,10 @@ Examples:
 
     would first search for the outgoing of the node with *uuid* starting with "a67fba41" and then select only those nodes of full_type *data.dict.Dict.|*.
+
+.. |Computer| replace:: :py:class:`Computer <aiida.orm.Computer>`
+.. |Node| replace:: :py:class:`Node <aiida.orm.Node>`
+.. |ProcessNode| replace:: :py:class:`ProcessNode <aiida.orm.ProcessNode>`
+.. |CalcJobNode| replace:: :py:class:`CalcJobNode <aiida.orm.CalcJobNode>`
+.. |User| replace:: :py:class:`User <aiida.orm.User>`
+.. |Group| replace:: :py:class:`Group <aiida.orm.Group>`
```
I thought we had to specify the complete Python dot-separated path to the classes. And why not use the more easily readable :py:class:`~aiida.orm.users.User`? The tilde (`~`) will make sure only `User` is written in the text.

ltalirz replied:
Agreed on both counts - fixed!

CasperWA reviewed (changes requested):
Cheers @ltalirz.

There are some sections that have been copied over that need a kind touch still I think. I've put in comments to highlight it.
Especially the lists of REST API field descriptions.

Also, personally, I prefer new lines to start with a capital letter, i.e., "Returns" instead of "returns".

ltalirz commented:
> There are some sections that have been copied over that need a kind touch still I think. I've put in comments to highlight it. Especially the lists of REST API field descriptions.

I guess you mean the tables in the section "Filtering results"?
I agree, I will work on making them look a bit nicer

> Also, personally, I prefer new lines to start with a capital letter, i.e., "Returns" instead of "returns".

Done!

CasperWA commented on docs/source/howto/share_data.rst:
```diff
@@ -167,6 +201,11 @@ For an extensive user documentation, see the :ref:`AiiDA REST API reference <ref
 
        $ curl https://aiida-dev.materialscloud.org/2dstructures/api/v4/users
 
+
+
```
```suggestion
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -176,15 +89,16 @@ Nodes
 
-        returns the list of *Node* objects. Every node object contains value of attribute called ``pbc1`` if present otherwise null.
+        Returns the list of |Node| objects.
+        Every node object contains value of attribute called ``pbc1`` if present otherwise null.
```
```suggestion
        Every node object contains value of attribute called ``pbc1`` if present otherwise ``null``.
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -758,18 +672,111 @@ CalcJobs
     Description:
 
-        returns a list of all the *User* objects.
+        Returns a list of all the *User* objects.
```
```suggestion
        Returns a list of all the |User| objects.
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -799,15 +806,15 @@ Users
     Description:
 
-        returns a lists of the *User* objects whose first name starts with ``"aii"``, regardless the case of the characters.
+        Returns a lists of the *User* objects whose first name starts with ``"aii"``, regardless the case of the characters.
```
```suggestion
        Returns a lists of the |User| objects whose first name starts with ``"aii"``, regardless the case of the characters.
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -834,15 +841,15 @@ Users
     Description:
 
-        returns the list of ten *Group* objects (``limit=10``) starting from the 1st row of the database table (``offset=0``) and the list will be ordered by ``user_id`` in descending order.
+        Returns the list of ten *Group* objects (``limit=10``) starting from the 1st row of the database table (``offset=0``) and the list will be ordered by ``user_id`` in descending order.
```
```suggestion
        Returns the list of ten |Group| objects (``limit=10``) starting from the 1st row of the database table (``offset=0``) and the list will be ordered by ``user_id`` in descending order.
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -886,7 +893,7 @@ Groups
     Description:
 
-        returns the details of the *Group* object with ``uuid="a6e5b..."``.
+        Returns the details of the *Group* object with ``uuid="a6e5b..."``.
```
```suggestion
        Returns the details of the |Group| object with ``uuid="a6e5b..."``.
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -913,31 +920,62 @@ Groups
+            <\http://localhost:5000/.../page/5?... >; rel=next,
+            <\http://localhost:5000/.../page/8?... >; rel=last
+
+
```
```suggestion
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -1142,3 +1178,10 @@ Examples:
+.. |ProcessNode| replace:: :py:class:`~aiida.orm.nodes.process.ProcessNode`
+.. |CalcJobNode| replace:: :py:class:`~aiida.orm.nodes.process.CalcJobNode`
+.. |User| replace:: :py:class:`~aiida.orm.users.User`
+.. |Group| replace:: :py:class:`Group <aiida.orm.Group>`
```
```suggestion
.. |Group| replace:: :py:class:`~aiida.orm.groups.Group`
```

CasperWA reviewed (changes requested):
Should be good to go after these minor changes.

ltalirz commented:
Thanks! 

I've added your suggestions + I fixed all tables in the "Filtering results" subsection so that they look nice and are easier to edit in the future + added sub sections to make that section more easy to follow.

CasperWA commented on docs/source/howto/share_data.rst:
```diff
@@ -155,9 +155,43 @@ Like all ``verdi`` commands, you can select a different AiiDA profile via the ``
+
+   Querying the base URL returns a list of all available endpoints.
+
+2. The *path* defining the requested *resource*, optionally followed by a more specific *endpoint*. For example::
```
```suggestion
2. The *path* defining the requested *resource*, optionally followed by a more specific *endpoint*.
   For example::
```

CasperWA commented on docs/source/howto/share_data.rst:
```diff
@@ -155,9 +155,43 @@ Like all ``verdi`` commands, you can select a different AiiDA profile via the ``
+       In order to query by *PK* you need to use the ``id`` filter (see below).
+       This also applies to :py:class:`~aiida.orm.users.User` s, which don't have UUIDs (but instead uses email).
+
+3. (Optional) The *query string* for filtering, ordering and pagination of results. For example::
```
```suggestion
3. (Optional) The *query string* for filtering, ordering and pagination of results.
   For example::
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -946,49 +995,246 @@ All of them must be followed by the operator ``=``.
+      - A comma-separated list of attributes to return.
+        Use together with ``attributes=true``.
+        Available in the endpoints ``/contents/attributes`` and ``/nodes``.
         Example:
 
         ::
```
```suggestion
        Example::
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -946,49 +995,246 @@ All of them must be followed by the operator ``=``.
+    * - ``tree_out_limit``
+      - specifies the limit on tree outgoing nodes.
+
+
```
```suggestion
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -946,49 +995,246 @@ All of them must be followed by the operator ``=``.
+
+
+Regular filters can be compounded, requiring all specified filters to apply.
+
```
```suggestion
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -946,49 +995,246 @@ All of them must be followed by the operator ``=``.
+      - users
+    * - ``label``
+      - string
+      - groups, nodes,
```
```suggestion
      - groups, nodes
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -946,49 +995,246 @@ All of them must be followed by the operator ``=``.
+Regular filters can be compounded, requiring all specified filters to apply.
+
+
+.. list-table:: Regular filters
```
Could you sort these (and the one above) alphabetically perhaps?

ltalirz replied:
done!

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -946,49 +995,246 @@ All of them must be followed by the operator ``=``.
 
-    :download_format: to specify download format in ``/download`` endpoint.
+    - ``node_type="data.code.Code."`` selects only objects of type |Code|
+    - ``node_type="data.remote.RemoteData."`` selects only objects of type *RemoteData*
```
```suggestion
    - ``node_type="data.remote.RemoteData."`` selects only objects of type :py:class:`~aiida.orm.nodes.data.remote.RemoteData`.
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -946,49 +995,246 @@ All of them must be followed by the operator ``=``.
+    Examples:
 
-    :download_format: to specify download format in ``/download`` endpoint.
+    - ``node_type="data.code.Code."`` selects only objects of type |Code|
```
```suggestion
    - ``node_type="data.code.Code."`` selects only objects of type |Code|.
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -946,49 +995,246 @@ All of them must be followed by the operator ``=``.
 
-    :download: in ``/download`` endpoint, if ``download=false`` it displays the content in the browser instead of downloading a file.
+.. note:: When using the *links/incoming* (*links/outgoing*) endpoints in combination with one or more filters, the filters are applied to the incoming (outgoing) nodes of the selected *id*.
+    For example, the request:
+
+        ::
```
```suggestion
    For example, the request::
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -946,49 +995,246 @@ All of them must be followed by the operator ``=``.
+    - ``http://localhost:5000/api/v4/nodes?id>578`` selects the nodes having an id larger than 578.
+    - ``http://localhost:5000/api/v4/users/?last_name<="m"`` selects only the users whose last name begins with a character in the range [a-m].
+
+
+
```
```suggestion
```

CasperWA commented on docs/source/reference/rest_api.rst:
```diff
@@ -946,49 +995,246 @@ All of them must be followed by the operator ``=``.
+
+Filter values should be specified as follows:
+
+.. list-table:: Regular filters
```
Maybe sort this list alphabetically as well?

ltalirz replied:
done!

CasperWA reviewed (approved):
One list was still not sorted alphabetically, but I don't want that to hold these changes back, since it's definitely not necessary.

Thanks @ltalirz ! I will squash+merge after merging in `develop`.
